### PR TITLE
Add custom attribute lineage extractor and Custom dispatcher

### PR DESCRIPTION
**Pull request by vidma**
* Add support for adding a `custom attribute lineage extractor`, fixes https://github.com/AbsaOSS/spline/issues/464
  - for the moment this is made very generic, so people can plug in their own implementation (returning `Option[Any]`)
  - later spline could provide its own default (hopefully, extensible with external additions)
  - a partial solution for #114 
* Add support for a `Custom dispatcher`, fixes https://github.com/AbsaOSS/spline/issues/463

@wajda

### CODE REVIEW

File: spark/agent/src/main/scala/za/co/absa/spline/harvester/conf/DefaultSplineConfigurer.scala [hunk 1]
```diff
@@ -41,8 +42,21 @@ object DefaultSplineConfigurer {
       */
     val MODE = "spline.mode"
     val MODE_DEFAULT = BEST_EFFORT.toString
-  }
 
+    /**
+     * Which lineage dispatcher should be used to report lineages (to Spline or elsewhere)
+     */
+    val LINEAGE_DISPATCHER_CLASS = "spline.lineage_dispatcher.className"
```
Comment (wajda) [inline spark/agent/src/main/scala/za/co/absa/spline/harvester/conf/DefaultSplineConfigurer.scala]: Codacy Issue found: [Field name 'LINEAGE_DISPATCHER_CLASS' does not match the regular expression '^[a-z][A-Za-z0-9]*$'](https://app.codacy.com/app/ABSA_OSS/spline/pullRequest?prid=4506530)

File: spark/agent/src/main/scala/za/co/absa/spline/harvester/conf/DefaultSplineConfigurer.scala [hunk 2]
```diff
@@ -65,11 +79,34 @@ class DefaultSplineConfigurer(configuration: Configuration, sparkSession: SparkS
...
+          .getConstructor(classOf[Configuration])
+          .newInstance(configuration)
+          .asInstanceOf[LineageDispatcher]
+    }
+  }
+
+  override lazy val attributeLineageExtractor: AbstractAttributeLineageExtractor = {
+    configuration.getString(LINEAGE_EXCTRACTOR_CLASS, LINEAGE_EXCTRACTOR_CLASS_DEFAULT) match {
+      case LINEAGE_EXCTRACTOR_CLASS_DEFAULT => new DisabledAttributeLineageExtractor
+      case className =>
```
Comment (wajda) [inline spark/agent/src/main/scala/za/co/absa/spline/harvester/conf/DefaultSplineConfigurer.scala]: Codacy Issue found: [Lower case pattern matching.](https://app.codacy.com/app/ABSA_OSS/spline/pullRequest?prid=4506530)

File: spark/agent/src/main/scala/za/co/absa/spline/harvester/conf/DefaultSplineConfigurer.scala [hunk 3]
```diff
@@ -41,8 +42,21 @@ object DefaultSplineConfigurer {
       */
     val MODE = "spline.mode"
     val MODE_DEFAULT = BEST_EFFORT.toString
-  }
 
+    /**
+     * Which lineage dispatcher should be used to report lineages (to Spline or elsewhere)
+     */
+    val LINEAGE_DISPATCHER_CLASS = "spline.lineage_dispatcher.className"
+    val LINEAGE_DISPATCHER_CLASS_DEFAULT = classOf[HttpLineageDispatcher].getCanonicalName
```
Comment (wajda) [inline spark/agent/src/main/scala/za/co/absa/spline/harvester/conf/DefaultSplineConfigurer.scala]: Codacy Issue found: [Field name 'LINEAGE_DISPATCHER_CLASS_DEFAULT' does not match the regular expression '^[a-z][A-Za-z0-9]*$'](https://app.codacy.com/app/ABSA_OSS/spline/pullRequest?prid=4506530)

**Comment by vidma** [comment]
looks like automatic `codacy review` not quite relevant. at least existing code nearby do not follow it.

File: spark/agent/src/main/scala/za/co/absa/spline/harvester/conf/DefaultSplineConfigurer.scala (hunk 3 above)
Comment (wajda) [inline spark/agent/src/main/scala/za/co/absa/spline/harvester/conf/DefaultSplineConfigurer.scala]: Codacy Issue found: [Field name 'LINEAGE_DISPATCHER_CLASS_DEFAULT' does not match the regular expression '^[a-z][A-Za-z0-9]*$'](https://app.codacy.com/app/ABSA_OSS/spline/pullRequest?prid=4506530)

File: spark/agent/src/main/scala/za/co/absa/spline/harvester/conf/DefaultSplineConfigurer.scala (hunk 2 above)
Comment (wajda) [inline spark/agent/src/main/scala/za/co/absa/spline/harvester/conf/DefaultSplineConfigurer.scala]: Codacy Issue found: [Lower case pattern matching.](https://app.codacy.com/app/ABSA_OSS/spline/pullRequest?prid=4506530)

File: spark/agent/src/main/scala/za/co/absa/spline/harvester/conf/DefaultSplineConfigurer.scala (hunk 1 above)
Comment (wajda) [inline spark/agent/src/main/scala/za/co/absa/spline/harvester/conf/DefaultSplineConfigurer.scala]: Codacy Issue found: [Field name 'LINEAGE_DISPATCHER_CLASS' does not match the regular expression '^[a-z][A-Za-z0-9]*$'](https://app.codacy.com/app/ABSA_OSS/spline/pullRequest?prid=4506530)

File: spark/agent/src/main/scala/za/co/absa/spline/harvester/conf/DefaultSplineConfigurer.scala [hunk 4]
```diff
@@ -41,8 +42,21 @@ object DefaultSplineConfigurer {
...
+
+    /**
+     * How Spline attribute lineage should behave.
+     *
+     * @see [[SplineAttributeLineageMode]]
+     */
+    val LINEAGE_EXCTRACTOR_CLASS = "spline.attribute_lineage_extractor.className"
```
Comment (wajda) [inline spark/agent/src/main/scala/za/co/absa/spline/harvester/conf/DefaultSplineConfigurer.scala]: Codacy Issue found: [Field name 'LINEAGE_EXCTRACTOR_CLASS' does not match the regular expression '^[a-z][A-Za-z0-9]*$'](https://app.codacy.com/app/ABSA_OSS/spline/pullRequest?prid=4506530)

File: spark/agent/src/main/scala/za/co/absa/spline/harvester/conf/DefaultSplineConfigurer.scala [hunk 5]
```diff
@@ -41,8 +42,21 @@ object DefaultSplineConfigurer {
...
+    val LINEAGE_EXCTRACTOR_CLASS_DEFAULT = classOf[DisabledAttributeLineageExtractor].getCanonicalName
```
Comment (wajda) [inline spark/agent/src/main/scala/za/co/absa/spline/harvester/conf/DefaultSplineConfigurer.scala]: Codacy Issue found: [Field name 'LINEAGE_EXCTRACTOR_CLASS_DEFAULT' does not match the regular expression '^[a-z][A-Za-z0-9]*$'](https://app.codacy.com/app/ABSA_OSS/spline/pullRequest?prid=4506530)

**Comment by wajda** [comment]
Hi @vidma, thank you very much for your PR. We'll take a look at it shortly. Could you please create an issue (if you didn't already) describing the goal and connect the PR to it?  
We don't add any new features into 0.4, only final bugfixes and polishing. But I'll happy to consider it for 0.5 though.

Re codacy, well.. some things could be irrelevant, depends on the context. In general we try to keep the config in a balanced shape and pay attention to codacy comments as much as possible and as long it makes sense. Let me review it and comment on each.

**Comment by vidma** [comment]
P.S. actually, earlier spline allowed to use a custom dispatcher, so in a sense, it's a fix for regression basically ;)

**Comment by wajda** [comment]
Fare enough :) We're also considering extracting Spline Spark Agent from the rest of Spline making it a separate project. It will allow adding extension points for integration with other tools easier. See #420

File: spark/agent/src/main/scala/za/co/absa/spline/harvester/conf/DefaultSplineConfigurer.scala (hunk 3 above)
Comment (wajda) [inline spark/agent/src/main/scala/za/co/absa/spline/harvester/conf/DefaultSplineConfigurer.scala]: Codacy Issue found: [Field name 'LINEAGE_DISPATCHER_CLASS_DEFAULT' does not match the regular expression '^[a-z][A-Za-z0-9]*$'](https://app.codacy.com/app/ABSA_OSS/spline/pullRequest?prid=4506530)

File: spark/agent/src/main/scala/za/co/absa/spline/harvester/conf/DefaultSplineConfigurer.scala (hunk 5 above)
Comment (wajda) [inline spark/agent/src/main/scala/za/co/absa/spline/harvester/conf/DefaultSplineConfigurer.scala]: Codacy Issue found: [Field name 'LINEAGE_EXCTRACTOR_CLASS_DEFAULT' does not match the regular expression '^[a-z][A-Za-z0-9]*$'](https://app.codacy.com/app/ABSA_OSS/spline/pullRequest?prid=4506530)

File: spark/agent/src/main/scala/za/co/absa/spline/harvester/conf/DefaultSplineConfigurer.scala (hunk 1 above)
Comment (wajda) [inline spark/agent/src/main/scala/za/co/absa/spline/harvester/conf/DefaultSplineConfigurer.scala]: Codacy Issue found: [Field name 'LINEAGE_DISPATCHER_CLASS' does not match the regular expression '^[a-z][A-Za-z0-9]*$'](https://app.codacy.com/app/ABSA_OSS/spline/pullRequest?prid=4506530)

File: spark/agent/src/main/scala/za/co/absa/spline/harvester/conf/DefaultSplineConfigurer.scala (hunk 4 above)
Comment (wajda) [inline spark/agent/src/main/scala/za/co/absa/spline/harvester/conf/DefaultSplineConfigurer.scala]: Codacy Issue found: [Field name 'LINEAGE_EXCTRACTOR_CLASS' does not match the regular expression '^[a-z][A-Za-z0-9]*$'](https://app.codacy.com/app/ABSA_OSS/spline/pullRequest?prid=4506530)

File: spark/agent/src/main/scala/za/co/absa/spline/harvester/conf/DefaultSplineConfigurer.scala (hunk 2 above)
Comment (wajda) [inline spark/agent/src/main/scala/za/co/absa/spline/harvester/conf/DefaultSplineConfigurer.scala]: Codacy Issue found: [Lower case pattern matching.](https://app.codacy.com/app/ABSA_OSS/spline/pullRequest?prid=4506530)

**Comment by wajda** [comment]
Hi @vidma, from what I see is that this PR addresses two different issues at once. Generally we prefer to address only one thing at a time. Could you please split it, so we can approach each one independently?
As for the the custom dispatcher, I agree it makes perfect sense. I'd be happy to review it as a separate PR.
Regarding the attribute extractor, I don't see any benefit in it, at least what is included in the PR. Yes, the attribute lineage is on our roadmap, but our implementation will most likely be different from yours, and thus I see no benefit in accepting your abstractions only, with no impl.

**Comment by vidma** [comment]
ok, I'll split the PRs.

> Yes, the attribute lineage is on our roadmap, but our implementation will most likely be different from yours, and thus I see no benefit in accepting your abstractions only, with no impl.

well, it allows people to easily plug in their own implementations or extra customizations/logic (in this can the attribute name can be more generic `ExecutionPlanExtra.AttributeLineage -> ExecutionPlanExtra.CustomData`), so it's super useful.

**Comment by vidma** [comment]
P.S. our existing attribute lineage impl works pretty well already, but I would need to check if we could share it (open source it or publish it as a jar as separate lib), and under what the license it would be - there is some possibility, but no promises yet. this depends on your interest too :)

@wajda

File: spark/agent/src/main/scala/za/co/absa/spline/harvester/conf/DefaultSplineConfigurer.scala [hunk 6]
```diff
@@ -65,11 +79,34 @@ class DefaultSplineConfigurer(configuration: Configuration, sparkSession: SparkS
...
+      case className =>
+        log debug s"Instantiating an AttributeLineageExtractor for class name: $className"
+        Class.forName(className.trim)
+          .getConstructor()
+          .newInstance()
+          .asInstanceOf[AbstractAttributeLineageExtractor]
+    }
+  }
 
   private lazy val harvesterFactory = new LineageHarvesterFactory(
```
Comment (vidma) [inline spark/agent/src/main/scala/za/co/absa/spline/harvester/conf/DefaultSplineConfigurer.scala]: alternatively if `harvesterFactory` or `harvester` class was configurable, we could remove explicit attribute lineage, and allow customizations to be added by creating an extended LineageHarvester(Factory) which wraps existing `spline` code. 
looks like the cleanest way to go for now...
@wajda

**Comment by vidma** [comment]
yes, a custom `SplineConfigurer` should be good for now.

**Pull request closed without merging**